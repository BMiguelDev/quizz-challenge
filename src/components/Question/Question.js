import React, { useEffect, useState } from 'react'
import { nanoid } from "nanoid"
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons'
import propTypes from "prop-types"

import styles from './Question.module.scss';


/**
 * Element containing a question and its possible answers
 */
export default function Question(props) {

  // Variable that holds an array with the answers scrambled (random order)
  const [scrambledAnswers, setScrambledAnswers] = useState([]);

  // Function that decodes strange characters in the question string
  function decodeQuestionString(string) {
    var text = document.createElement("textarea");
    text.innerHTML = string;
    return text.value;
  }

  // On first component mount, create array with all answers in a scrambled order and store it in <scrambledAnswers> variable
  useEffect(() => {
    if (scrambledAnswers.length === 0) {
      const newAnswersArray = props.questionData.incorrect_answers.slice();
      newAnswersArray.push(props.questionData.correct_answer);
      setScrambledAnswers(newAnswersArray.sort((a, b) => 0.5 - Math.random()));
    }
  }, []); // eslint-disable-line react-hooks/exhaustive-deps


  // Function that decides className property of answer for correct coloring when quizz ends
  function getAnswerClassName(answer) {
    const isShowResults = props.isShowResults;
    if (isShowResults) {
      return answer === props.questionData.correct_answer ?
        (answer === props.questionData.answerSelected ? `${styles.answer} ${styles.answer_selected} ${styles.answer_correct}` : `${styles.answer} ${styles.answer_correct}`) :
        (answer === props.questionData.answerSelected ? `${styles.answer} ${styles.answer_incorrect} ${styles.answer_selected}` : `${styles.answer} ${styles.answer_dimmed}`)

    } else {
      return props.questionData.answerSelected === answer ? `${styles.answer} ${styles.answer_selected}` : styles.answer
    }
  }


  return (
    <div className={styles.question_container}>
      <h3 className={styles.question}>{decodeQuestionString(props.questionData.question)}</h3>
      <div className={styles.answers_container}>
        {
          scrambledAnswers.map(answer => (
            <div key={nanoid()} className={styles.answer_container}>
              <p
                className={getAnswerClassName(answer)}
                onClick={() => props.handleSelectAnswer(props.questionIndex, answer)}
                aria-label="answer"
              >
                {decodeQuestionString(answer)}
              </p>
              {getAnswerClassName(answer)===`${styles.answer} ${styles.answer_selected} ${styles.answer_correct}` && <FontAwesomeIcon icon={faCheck} /> /*<i className="fa-solid fa-check"></i>*/}
            </div>
          ))
        }
      </div>
      <hr className={styles.question_separator} />
    </div>
  )
}

Question.propTypes = {
  /**
   * Question information, including:
   * - category
   * - type
   * - difficulty
   * - question text
   * - correct answer
   * - array of incorrect answers
   * - selected answer (optional)
   */
  questionData: propTypes.shape({
    category: propTypes.string,
    type: propTypes.string,
    difficulty: propTypes.string,
    question: propTypes.string,
    correct_answer: propTypes.string,
    incorrect_answers: propTypes.arrayOf(propTypes.string),
    answerSelected: propTypes.string
  }),
  /**
   * Id of the question
   */
  questionIndex: propTypes.number,
  /**
   * Click handler for the answers
   */
  handleSelectAnswer: propTypes.func,
  /**
   * Flag indicating whether or not to show the question's correct answer
   */
  isShowResults: propTypes.bool
}
