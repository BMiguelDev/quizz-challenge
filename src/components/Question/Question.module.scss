// Question content
.question_container {
    width: 100%;
    padding: .75rem 1rem .25rem 1rem;
    box-shadow: 0 -.175rem .45rem .005rem var(--question-box-shadow-color);
    border-radius: 1rem;

    .question {
        color: var(--question-color);
        margin: 0;
        font-size: 1.075rem;
        transition: all .2s ease-in-out;
    }

    .answers_container {
        display: flex;
        align-items: center;
        // gap: 1.25rem;
        gap: 2.5%;
        justify-content: flex-start;
        padding: 0 .35rem;

        .answer_container {
            display: flex;
            position: relative;

            svg {
                position: absolute;
                right: -10%;
                top: 0;
                font-size: 1.75rem;
                color: var(--answer-checkmark-color);
                opacity: 0;
                filter: drop-shadow(0 0 3px rgb(228, 255, 205));
                z-index: 3;

                @keyframes checkmark-opacity {
                    0% {
                        opacity: 0;
                    }

                    100% {
                        opacity: 1;
                    }
                }

                @keyframes checkmark-glow {
                    0% {
                        filter: drop-shadow(0 0 0 rgb(228, 255, 205));
                    }

                    100% {
                        filter: drop-shadow(0 0 1rem rgb(80, 229, 0));
                    }
                }

                animation: checkmark-glow 1s ease .8s infinite alternate,
                checkmark-opacity 1s ease .5s;
                animation-fill-mode: forwards;
            }

            .answer {
                border: .1rem solid var(--answer-border-color);
                border-radius: 1rem;
                padding: .3rem 1.25rem;
                cursor: pointer;
                font-size: .75rem;
                font-weight: 600;
                color: var(--answer-color);
                transform: scale(1);
                transition: all .2s ease-in-out;
                margin: .75rem 0 .35rem 0;
                line-height: .825rem;

                &:hover:not(.answer_incorrect, .answer_dimmed, .answer_correct, .answer_selected) {
                    transform: scale(1.15);
                }

                &.answer_selected {
                    background-color: var(--answer-selected-color);
                    transform: scale(1.15);
                }

                &.answer_correct {
                    background-color: var(--answer-correct-color);
                    @keyframes answer-correct-animation {
                        0% {
                            transform: scale(1);
                        }

                        100% {
                            transform: scale(1.2);
                        }
                    }

                    animation: answer-correct-animation 1.5s ease;
                    animation-fill-mode: forwards;
                    z-index: 2;
                }

                &.answer_selected.answer_correct {
                    background-color: var(--answer-correct-selected-color);
                    color: var(--answer-correct-selected-text-color);
                }

                &.answer_incorrect {
                    background-color: var(--answer-incorrect-color);
                    color: var(--answer-dimmed-color-transparent);
                }

                &.answer_dimmed {
                    color: var(--answer-dimmed-color);
                    opacity: .7;
                }
            }
        }
    }

    .question_separator {
        border: .01rem solid var(--question-separator-border-color);
        transition: all .2s ease-in-out;
    }
}


@import './media';
