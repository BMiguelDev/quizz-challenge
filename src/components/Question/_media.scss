// Media query for mobile extremelly small screens
@media screen and (max-width: 400px),
screen and (max-height: 450px) {

    // Question content
    .question_container {

        .question {
            font-size: .7rem;
        }

        .answers_container {
            width: 100%;
            gap: unset;
            justify-content: space-evenly;

            .answer_container {
                width: 20%;

                svg {
                    top: 1%;
                    font-size: 1.5rem;
                }

                .answer {
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    text-align: center;
                    width: 100%;
                    word-break: break-word;
                    padding: .15rem .4rem;
                    font-size: .525rem;
                    margin: .35rem 0 .25rem 0;

                    &.answer_correct {

                        @keyframes answer-correct-animation {
                            0% {
                                transform: scale(1);
                            }

                            100% {
                                transform: scale(1.15);
                            }
                        }
                    }
                }
            }
        }

        .question_separator {
            width: 100%;
            margin: 0;
            margin-top: .15rem;
        }
    }
}


// Media query for mobile content (portrait)
@media screen and (max-width: 550px) and (min-height: 450px) {

    // Question content
    .question_container {

        .question {
            font-size: .875rem;
        }

        .answers_container {
            width: 100%;
            gap: unset;
            justify-content: space-evenly;

            .answer_container {
                width: 20%;

                svg {
                    top: 1%;
                    font-size: 1.5rem;
                }

                .answer {
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    text-align: center;
                    width: 100%;
                    word-break: break-word;
                    padding: .15rem .4rem;
                    font-size: .625rem;
                }
            }
        }

        .question_separator {
            width: 100%;
            margin: 0;
            margin-top: .25rem;
        }
    }
}


// Media query for mobile content (landscape)
@media screen and (max-height: 1200px) and (min-width: 550px) and (max-width: 900px),
screen and (max-height: 750px) and (min-width: 900px) {

    // Question content
    .question_container {

        .question {
            font-size: .875rem;
        }

        .answers_container {
            gap: 1.5rem;
            justify-content: center;

            .answer_container {
                width: 20%;

                .answer {
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    text-align: center;
                    width: 100%;
                    font-size: .675rem;
                    word-break: break-word;
                    padding: .2rem .75rem;

                    &.answer_correct {

                        @keyframes answer-correct-animation {
                            0% {
                                transform: scale(1);
                            }
    
                            100% {
                                transform: scale(1.175);
                            }
                        }
                    }
                }
            }
        }

        .question_separator {
            margin-top: 0rem;
            margin-bottom: .2rem;
        }
    }
}