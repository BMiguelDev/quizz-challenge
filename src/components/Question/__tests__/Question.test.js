import { cleanup, render, screen } from "@testing-library/react"
import renderer from 'react-test-renderer';

import Question from "../Question";


// Mock the nanoid external module used in Question component
jest.mock("nanoid", () => ({
    __esModule: true,
    nanoid: () => Date.now()
}))


afterEach(() => {
    cleanup();
})


test('number of answers for multiple choice question should be 4', () => {
    const mockFunction = jest.fn();

    const questionDataTest = {
        category: "Entertainment: Video Games",
        correct_answer: "The Zin Empire",
        difficulty: "easy",
        incorrect_answers: ['The Brotherhood', 'The Ronin', 'The Sons of Samedi'],
        question: "Which of these is NOT the name of a rival gang in the video game Saint&#039;s Row 2?",
        type: "multiple"
    }

    render(<Question questionData={questionDataTest} questionIndex={0} handleSelectAnswer={mockFunction} isShowResults={false} />);

    const answerElements = screen.getAllByLabelText(/answer/i);

    expect(answerElements).toHaveLength(4);
})

test('number of answers for true/false question should be 2', () => {
    const questionDataTest = {
        category: "Science: Computers",
        correct_answer: "False",
        difficulty: "medium",
        incorrect_answers: ['True'],
        question: "The common software-programming acronym &quot;I18N&quot; comes from the term &quot;Interlocalization&quot;.",
        type: "boolean"
    }

    const mockFunction = jest.fn();

    render(<Question questionData={questionDataTest} questionIndex={0} handleSelectAnswer={mockFunction} isShowResults={false} />);

    const answerElements = screen.getAllByLabelText(/answer/i);

    expect(answerElements).toHaveLength(2);
})

test("component should always match snapshot", () => {
    const mockFunction = jest.fn();
    const questionDataTest = {
        category: "Entertainment: Video Games",
        correct_answer: "The Zin Empire",
        difficulty: "easy",
        incorrect_answers: ['The Brotherhood', 'The Ronin', 'The Sons of Samedi'],
        question: "Which of these is NOT the name of a rival gang in the video game Saint&#039;s Row 2?",
        type: "multiple"
    }

    const tree = renderer.create(<Question questionData={questionDataTest} questionIndex={0} handleSelectAnswer={mockFunction} isShowResults={false} />);

    expect(tree).toMatchSnapshot();
});
