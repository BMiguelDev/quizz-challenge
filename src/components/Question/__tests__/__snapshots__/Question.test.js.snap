// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component should always match snapshot 1`] = `
<div
  className="question_container"
>
  <h3
    className="question"
  >
    Which of these is NOT the name of a rival gang in the video game Saint's Row 2?
  </h3>
  <div
    className="answers_container"
  />
  <hr
    className="question_separator"
  />
</div>
`;
